[JSC] Clean up and enhance X86 vector instructions

Constellation opened the pull request:
#### 788a24f32f108ced27e623bb590d2f1ddb4eac3a
<pre>
[JSC] Clean up and enhance X86 vector instructions
<a href="https://bugs.webkit.org/show_bug.cgi?id=248729">https://bugs.webkit.org/show_bug.cgi?id=248729</a>
rdar://102942709

Reviewed by Justin Michaud.

This patch fixes X86Assembler&apos;s naming convention. Plus, this patch enhances splat / extract implementation to use
more efficient instructions. Also we check SSE3 status since some of used instructions are only available in SSE3,
which is not mandatory in x64.

* Source/JavaScriptCore/assembler/MacroAssemblerX86Common.cpp:
(JSC::MacroAssemblerX86Common::collectCPUFeatures):
* Source/JavaScriptCore/assembler/MacroAssemblerX86Common.h:
(JSC::MacroAssemblerX86Common::supportsSSE3):
(JSC::MacroAssemblerX86Common::supportsSupplementalSSE3):
* Source/JavaScriptCore/assembler/MacroAssemblerX86_64.h:
(JSC::MacroAssemblerX86_64::vectorExtractLane):
(JSC::MacroAssemblerX86_64::vectorSplat):
* Source/JavaScriptCore/assembler/X86Assembler.h:
(JSC::X86Assembler::vextractps):
(JSC::X86Assembler::pshufd_rr):
(JSC::X86Assembler::pshufb_rr):
(JSC::X86Assembler::pshuflw_rr):
(JSC::X86Assembler::pshufhw_rr):
(JSC::X86Assembler::punpcklqdq_rr):
(JSC::X86Assembler::shufps_rr):
(JSC::X86Assembler::shufpd_rr):
(JSC::X86Assembler::movddup_rr):
(JSC::X86Assembler::movhlps_rr):
(JSC::X86Assembler::movshdup_rr):
(JSC::X86Assembler::movsldup_rr):
(JSC::X86Assembler::pshufd): Deleted.
(JSC::X86Assembler::pshufb): Deleted.
(JSC::X86Assembler::pshuflw): Deleted.
(JSC::X86Assembler::pshufhw): Deleted.
(JSC::X86Assembler::punpcklqdq): Deleted.
(JSC::X86Assembler::shufps): Deleted.
(JSC::X86Assembler::shufpd): Deleted.
(JSC::X86Assembler::movddup): Deleted.

Canonical link: <a href="https://commits.webkit.org/257381@main">https://commits.webkit.org/257381@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b581cdd5aafce301f8b25b21eded2379714340c4

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/98733 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/7944 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/31875 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/108153 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/168406 "Built successfully and passed tests") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/8474 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/85312 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/91264 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/106071 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/104415 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/6415 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/89987 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/33403 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/88241 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/21330 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/76329 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/89499 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/1865 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/22859 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/85282 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/1773 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/45330 "Found 1 new test failure: media/video-inactive-playback.html (failure)") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/28758 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/6719 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/42314 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/88135 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/2563 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/3166 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/19732 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

hyjorc1 commented on Source/JavaScriptCore/assembler/MacroAssemblerX86_64.h:
```diff
@@ -2158,18 +2158,37 @@ class MacroAssemblerX86_64 : public MacroAssemblerX86Common {
+            else {
+                ASSERT(src == dest);
+                m_assembler.shufps_rr(0, dest, dest);
+            }
```
I think it should have imm8 `3` instead of `0`. So, it should
be
```
if (src != dest)
    m_assembler.pshufd_rr(lane.m_value, src, dest);
else {
    ASSERT(src == dest);
    m_assembler.shufps_rr(lane.m_value, dest, dest);
}
```

Constellation replied:
Nice, fixed.

justinmichaud commented on Source/JavaScriptCore/assembler/MacroAssemblerX86_64.h:
```diff
@@ -2158,18 +2158,37 @@ class MacroAssemblerX86_64 : public MacroAssemblerX86Common {
     void vectorExtractLane(SIMDLane simdLane, TrustedImm32 lane, FPRegisterID src, FPRegisterID dest)
     {
+        // For lane 0, we just move since we do not ensure the upper bits.
+        if (!lane.m_value) {
```
This seems wrong

Constellation replied:
Talked offline. I think this is correct. This is only for SIMDLane::f32x4 and SIMDLane::f64x2, and our definition is,
```
  1663 VectorExtractLaneFloat64 U:G:8, U:F:128, D:F:64
  1664     Imm, Tmp, Tmp
  1665 VectorExtractLaneFloat32 U:G:8, U:F:128, D:F:32
  1666     Imm, Tmp, Tmp
  ```

justinmichaud reviewed (approved):
r=me

Constellation committed:
[JSC] Clean up and enhance X86 vector instructions
https://bugs.webkit.org/show_bug.cgi?id=248729
rdar://102942709

Reviewed by Justin Michaud.

This patch fixes X86Assembler's naming convention. Plus, this patch enhances splat / extract implementation to use
more efficient instructions. Also we check SSE3 status since some of used instructions are only available in SSE3,
which is not mandatory in x64.

* Source/JavaScriptCore/assembler/MacroAssemblerX86Common.cpp:
(JSC::MacroAssemblerX86Common::collectCPUFeatures):
* Source/JavaScriptCore/assembler/MacroAssemblerX86Common.h:
(JSC::MacroAssemblerX86Common::supportsSSE3):
(JSC::MacroAssemblerX86Common::supportsSupplementalSSE3):
* Source/JavaScriptCore/assembler/MacroAssemblerX86_64.h:
(JSC::MacroAssemblerX86_64::vectorExtractLane):
(JSC::MacroAssemblerX86_64::vectorSplat):
* Source/JavaScriptCore/assembler/X86Assembler.h:
(JSC::X86Assembler::vextractps):
(JSC::X86Assembler::pshufd_rr):
(JSC::X86Assembler::pshufb_rr):
(JSC::X86Assembler::pshuflw_rr):
(JSC::X86Assembler::pshufhw_rr):
(JSC::X86Assembler::punpcklqdq_rr):
(JSC::X86Assembler::shufps_rr):
(JSC::X86Assembler::shufpd_rr):
(JSC::X86Assembler::movddup_rr):
(JSC::X86Assembler::movhlps_rr):
(JSC::X86Assembler::movshdup_rr):
(JSC::X86Assembler::movsldup_rr):
(JSC::X86Assembler::pshufd): Deleted.
(JSC::X86Assembler::pshufb): Deleted.
(JSC::X86Assembler::pshuflw): Deleted.
(JSC::X86Assembler::pshufhw): Deleted.
(JSC::X86Assembler::punpcklqdq): Deleted.
(JSC::X86Assembler::shufps): Deleted.
(JSC::X86Assembler::shufpd): Deleted.
(JSC::X86Assembler::movddup): Deleted.

Canonical link: https://commits.webkit.org/257381@main
